File: dp/10099.py
# bfs - AC
from math import ceil
import sys

dp = [[0 for column in range(106)] for row in range(106)]
t = 1
while True:	
	try:
		nodes, edges = [int(i) for i in sys.stdin.readline().split()]
	except:
		break
	
	if nodes == 0 and edges == 0: break
		
	for i in range(1, nodes+1):
		for j in range(1, nodes+1):
			dp[i][j] = -1e6
		dp[i][i] = 0

	for _ in range(edges):
		src, dest, weight = [int(i) for i in sys.stdin.readline().split()]
		dp[src][dest] = dp[dest][src] = weight

	src, dest, weight = [int(i) for i in sys.stdin.readline().split()]

	for k in range(1, nodes+1):
		for i in range(1, nodes+1):
			for j in range(1, nodes+1):
				dp[i][j] = dp[j][i] = max(dp[i][j], min(dp[i][k], dp[k][j]))

	result = ceil(weight/(dp[src][dest]-1))
	print('Scenario #%d' % t)
	print('Minimum Number of Trips = %d\n' % result)
	t += 1

File: dp/10099.cpp
#include <iostream>
#include <algorithm>
#include <cmath>

using namespace std;


int nodes, edges, src, dest, weight, tourist, result, t;
int cache[150][150];

int floyd(){
	for(int k=1; k<=nodes; k++){
		for(int i=1; i<=nodes; i++){
			for(int j=1; j<=nodes; j++){
				cache[i][j] = cache[j][i] = max(cache[i][j], min(cache[i][k], cache[k][j]));
			}
		}
	}
	return cache[src][dest]-1;
}


int main(int argc, char const *argv[]){
	t = 1;	
	while((cin >> nodes >> edges) && (nodes != 0)){

		for(int i=1; i<=nodes; i++){
			for(int j=1; j<=edges; j++){
				cache[i][j] = 0;
			}
		}

		for(int i=0; i<edges; i++){
			cin >> src >> dest >> weight;
			cache[src][dest] = weight;
			cache[dest][src] = weight;
		}
		cin >> src >> dest >> tourist;

		cout << "Scenario #" << t++ << endl;
		int out = floyd(); 
		out = (out != 0) ? ceil((tourist*1.0)/out) : 0;
		cout << "Minimum Number of Trips = " << out << endl << endl;

	}

	return 0;
}